src/client.rs

```diff
@@ -295,32 +311,57 @@ where
             Err(_) => return Err(Error::SocketError),
         };
 
-        // Compare server and client hashes.
-        let password_hash = md5_hash_password(&config.user.name, &config.user.password, &salt);
 
-        if password_hash != password_response {
-            debug!("Password authentication failed");
-            wrong_password(&mut write, &config.user.name).await?;
-            return Err(Error::ClientError);
+
+        let mut target_pool: ConnectionPool = ConnectionPool::default();
+        let mut transaction_mode = false;
+
+        if admin {
+            let correct_user = config.general.admin_username.as_str();
+            let correct_password = config.general.admin_password.as_str();
+
+            // Compare server and client hashes.
+            let password_hash = md5_hash_password(correct_user, correct_password, &salt);
+            if password_hash != password_response {
+                debug!("Password authentication failed");
+                wrong_password(&mut write, user).await?;
+                return Err(Error::ClientError);
+            }
+        } else {
+            target_pool = match get_pool(database.clone(), user.clone()) {
+                Some(pool) => pool,
+                None => {
+                    error_response(
+                        &mut write,
+                        &format!("No pool configured for database: {:?}, user: {:?}", database, user),
+                    ).await?;
+                    return Err(Error::ClientError)
+                }
+            };
+            transaction_mode = target_pool.settings.pool_mode == "transaction";
```

Comment:
This concept of `target_pool` simplifies the implementation. Once a client connects it is permanently assigned a `target_pool`. This `target_pool` connects to one cluster (with whatever shards/instances configured for that cluster) using a specific user. Almost everything after more or less behaves identically to how Pgcat operated before